src/devices/bus/hp_dio/hp_dio.h

```diff
@@ -105,13 +105,13 @@ class dio16_device : public device_t
 	void reset_in(int state);
 
 	// output lines
-	int irq1_out() const { return (m_irq[0] & ~m_bus_index) ? 1 : 0; }
-	int irq2_out() const { return (m_irq[1] & ~m_bus_index) ? 1 : 0; }
-	int irq3_out() const { return (m_irq[2] & ~m_bus_index) ? 1 : 0; }
-	int irq4_out() const { return (m_irq[3] & ~m_bus_index) ? 1 : 0; }
-	int irq5_out() const { return (m_irq[4] & ~m_bus_index) ? 1 : 0; }
-	int irq6_out() const { return (m_irq[5] & ~m_bus_index) ? 1 : 0; }
-	int irq7_out() const { return (m_irq[6] & ~m_bus_index) ? 1 : 0; }
+	int irq1_out() const { return m_irq[0] != 0; }
+	int irq2_out() const { return m_irq[1] != 0; }
+	int irq3_out() const { return m_irq[2] != 0; }
+	int irq4_out() const { return m_irq[3] != 0; }
+	int irq5_out() const { return m_irq[4] != 0; }
+	int irq6_out() const { return m_irq[5] != 0; }
+	int irq7_out() const { return m_irq[6] != 0; }
```

Comment:
I changed this for various reasons:
1. It clearly looks like unfinished code to me. Classes in `hp_dio.cpp` have a few things that make no sense but they have never been tested before because they were only used in machines that are marked as "not working" (in hp9k_3xx.cpp).
2. My understanding of how `dio16_device` was meant to work is as follows (any of the original authors can confirm, please?).
`dio16_device` uses 16-bit words for each interrupt line (`m_irq[]`). Every card has a bit reserved in these words (its index number which is handed out by returning `m_bus_index` and incrementing it each time a card is registered through `dio16_device::add_card`). In order to know whether a line is asserted or not m_irq[x] should be tested to be zero/non-zero. But instead that nonsensical `&~m_bus_index` is used. IRL interrupt lines are driven by open-collector outputs so that multiple cards can share them.
3. After my changes `dio16_device` works correctly in hp98x6 systems.